docs/programmable-control-flow.md

```diff
@@ -0,0 +1,145 @@
+# Programmable Control Flow
+
+New language features:
+
+* Lambda parameters can be values, which turns the lambda into a case.
+  * The parameter position takes a type that can be the right-hand side in an `==` operator with the pattern value. When the lambda executes, it first checks all such equalities, throwing `NoMatch` if any aren't satisfied.
+* Negation in types to indicate that some type won't be thrown.
+* Automatically convert an exception that we don't allow in the function signature to a local return value and type check against the return value.
+  * Or enclose all bodies with a try/catch that catches Return[$ResultType] and returns the value
+  * May be better to catch each Return[T] exception that's not in the `throws` signature and unwrap it
+* Throw `NoCatch` in a catch clause to backtrack
+  * A catch that doesn't execute anything rethrows the exception
+
+## Example Usage
+
+```ts
+// Because `f` can't throw Return[U64], the `return` below becomes a local
+// return of U64.
+f(iter: Iterator[U64], accum: U64): U64
+{
+  // try
+  // {
+  for iter
+  {
+    x =>
+    accum = accum + x;
+
+    match x
+    {
+      // non-local return, `for` returns 0
+      { 0 => return 0 }
+
+      // `for` should finish
+      { 1 => break }
+
+      // `for` should move to the next element
+      { 2 => continue }
+    }
+
+    // local return
+    accum = accum + 1
+  }
+
+  accum
+  // }
+  // catch
+  // {
+  //   { $0: Return[$ReturnType] => $0.value }
+  // }
+}
+
+if (a) { e1 } else case {b} { e2 } else { e3 }
+// (apply `else`
+//  (tuple
+//    [
+//      (apply `else`
+//        (tuple
+//          [
+//            (apply `if` (tuple [a (lambda e1)]))
+//            (apply `case` (tuple [(lambda b) (lambda e2)]))
+//          ]))
+//      (lambda e3)
+//    ]))
+
+```
+
+## Cases
+
+A case can can backtrack by throwing `NoMatch`.
+Labelled `NoMatch` gets multi-level backtrack.
+A match that doesn't match anything throws `NoMatch`.
+`match` can't be written in the source language because we want to detect exhaustive matches.
+
+> TODO: destructure in patterns
```

Comment:
FWIW, the sort of analogy in Haskell is View Patterns, which don't project a `Bool` like `==` here, but project an arbitrary value which is then subjected to the usual pattern handling (I think including additional views, though I'm not sure why you'd do that).  So you'll see a lot of `view :: ... -> Maybe ...` views that get used like `foo (view x -> Just z) = ... z ...`.  (see https://gitlab.haskell.org/ghc/ghc/-/wikis/view-patterns)

Reply:
I'm definitely not following this. I'll bug you for a walkthrough.